src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_expert_opt.cpp

```diff
@@ -0,0 +1,1215 @@
+        layout layout_down_out(ov::PartialShape{batch, static_cast<int>(config.hidden_size)}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_gateup_out, false);  // up
+        internal_buffers.emplace_back(layout_down_out, false);    // y
+        // onednn: scratch.x, scratch.routing_weights = gather(x, ...)
+        //         scratch.up = up(scratch.x)
+        //         scratch.gate = gate(scratch.x) * scratch.up
+        //         scratch.y = down(scratch.gate) * routing_weights
+        internal_buffers.emplace_back(layout_down_out, false);  // x, scratch.x has same layout with down output
+        layout routing_layout(ov::PartialShape{batch * max_topk}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_down_out, false);    // routing_weights
+        internal_buffers.emplace_back(layout_gateup_out, false);  // gate, scratch.gate has same layout with up
+        // expert masks for gpu
+        layout index_layout(ov::PartialShape{batch}, ov::element::i32, cldnn::format::bfyx);
+        for (int i = 0; i < expert_num; i++) {
+            internal_buffers.emplace_back(index_layout, true);  // batch
+            internal_buffers.emplace_back(index_layout, true);  // topk
+        }
+
+        return internal_buffers;
+    }
+
+    void prepare_internal_buffers(typed_primitive_inst<moe_expert>& instance, scratch_buffers& scratch, bool is_single_batch) {
+        const auto& intermediates_memories = instance.get_intermediates_memories();
+        scratch.topk_id = intermediates_memories[0];
+        scratch.topk_weights = intermediates_memories[1];
+        scratch.up = intermediates_memories[2];
+        scratch.y = intermediates_memories[3];
+        if (!is_single_batch) {
+            scratch.x = intermediates_memories[4];
+            scratch.routing_weights = intermediates_memories[5];
+            scratch.gate = intermediates_memories[6];
+            const auto& config = instance.get_typed_desc<moe_expert>()->_config;
+            int expert_num = static_cast<int>(config.expert_num);
+            scratch.expert_masks.resize(expert_num);
+            for (int i = 0; i < expert_num; i++) {
+                scratch.expert_masks[i].batch = intermediates_memories[7 + 2 * i + 0];
+                scratch.expert_masks[i].topk = intermediates_memories[7 + 2 * i + 1];
+            }
+        }
+    }
+
+    void get_expert_mask_from_gpu(const MOEExpert::Config& config, memory::ptr mem, stream& stream, expert_mask_cpu& expert_mask) {
+        // shape: [batch, topk]
+        auto layout = mem->get_layout();
+        const auto& shape = layout.get_shape();
+
+        int max_expert_num = static_cast<int>(config.expert_num), max_topk = static_cast<int>(config.topk), max_tokens = static_cast<int>(shape[0]);
+
+        expert_mask.pred_flag.resize(max_expert_num, 0);
```

Comment:
I think we can store these masks in the implementation to avoid reallocating them on every invocation

Reply:
It will bring memory reset to avoid data corruption.